site/public/ts/controllers/LoginController.ts

```diff
@@ -0,0 +1,151 @@
+        authState = undefined;
+        loginHeaderContainer.classList.remove("hide");
+        logoutHeaderButton.classList.add("hide");
+    }
+
+    const onLoginError = function(e: {error: string}): void {
+        console.error(e);
+    }
+
+    const setAuthState = async function(as: AuthState): Promise<void> {
+        if (!authState) {
+            authState = as;
+            return;
+        }
+        console.error("Attempt to load multiple auth states!");
+        await authState.logOut().then(() => authState = as);
+    }
+
+    const onGoogleLogin = async function(_: gapi.auth2.GoogleUser): Promise<void> {
+        await setAuthState(new GoogleAuthState());
+        onLogin();
+    }
+
+    const onNoAuthLogin = function(username: string): void {
+        setAuthState(new NoAuthState(username));
+        onLogin();
+    }
+
+    const onNoAuthSubmitted = function(): void {
+        const username = (<HTMLInputElement>document.getElementById("no-auth-user-input")).value;
+        if (username === "") {
+            alert("User Name must not be blank");
+            return;
+        }
+        onNoAuthLogin(username);
+    }
+
+    const toggle = function(): void {
+        loginPopup.classList.toggle("invisible");
+        overlay.classList.toggle("invisible");
+    }
+
+    return {
+        Init: async function(): Promise<number> {
+            isOpen = false;
+
+            loginHeaderButton.onclick = () => {
+                LoginController.Toggle();
+            };
+
+            logoutHeaderButton.onclick = async () => {
+                if (authState)
+                    await authState.logOut();
+                onLogout();
+            };
+
+            overlay.addEventListener("click", () => {
+                if (LoginController.IsOpen())
+                    LoginController.Toggle();
+            });
+
+            // Setup each auth method if they are loaded in the page
+            if (googleAuthMeta) {
+                // Load GAPI script
+                await LoadDynamicScript("https://apis.google.com/js/platform.js");
+
+                // Render sign in button
+                gapi.signin2.render('login-popup-google-signin', {
+                    'scope': 'profile email',
+                    'width': 240,
+                    'height': 50,
+                    'longtitle': true,
+                    'onsuccess': onGoogleLogin,
+                    'onfailure': onLoginError
+                });
+
+                // Load 'auth2' from GAPI and then initialize w/ meta-data
+                await new Promise<void>((resolve) => gapi.load('auth2', resolve));
+                await gapi.auth2.init(new class implements ClientConfig {
+                    // Written like this to make ESLint happy
+                    public client_id?: string = googleAuthMeta.getAttribute("content");
+                }).then(async (_) => {}); // Have to explicitly call .then
+
+            }
+            else if (noAuthMeta) {
```

Comment:
is there any reason in particular this is an `else if` instead of an `if?  It would be useful to use `if` so we can make sure it works with multiple auth methods at the same time.